canonical/traefik-k8s-operator
# Integrate with oathkeeper via `forward-auth` interface

## Issue

This PR adds integration with Oathkeeper - the policy decision point in Identity and Access Proxy architecture, by supporting a new `forwardAuth` middleware.

In order to integrate Traefik into IAP, it must update routing rules for the applications that requested to be protected with the `ForwardAuth` middleware. Its definition must include headers and oathkeeper's decisions address.

## Solution

This PR integrates traefik with oathkeeper with the use of `forward-auth` interface.
Oathkeeper provides Traefik with its decisions address, allowed return headers and application names that requested the IAP protection. This information is processed by Oathkeeper from `auth-proxy` relation.
Traefik updates the relation databag with app names it provides ingress for, so that Oathkeeper can compare the app names and evaluate if they can be protected using forwardAuth.
If an app is related with `ingress-per-app/leader/unit`, Traefik will update its routes and add the middleware.

## Context

[Traefik ForwardAuth upstream docs](https://doc.traefik.io/traefik/middlewares/http/forwardauth/)
[IAP interfaces design](https://docs.google.com/document/d/13a3OEQ_Q4Fwvv9F57wgBYJm4SUFQz4W5M32UVqye4Rg/edit?usp=sharing)

## Testing Instructions

1. Pack and deploy the charms.
As `forward-auth` relies on `auth-proxy` relation data, I'm using zinc-k8s from [this branch](https://github.com/natalian98/zinc-k8s-operator/tree/test-proxy-interfaces) as a dummy example to test that the relation data is then transferred to traefik.
Oathkeeper is built from [this](https://github.com/canonical/oathkeeper-operator/tree/IAM-500-forward-auth-relation-implementation) branch.
```
juju deploy ./traefik-k8s_ubuntu-20.04-amd64.charm traefik-ingress --trust --resource traefik-image=ghcr.io/canonical/traefik:2.10.4
juju deploy ./oathkeeper*.charm --resource oci-image=$(yq eval '.resources.oci-image.upstream-source' metadata.yaml) --trust
juju deploy ./zinc-k8s*.charm --resource zinc-image=ghcr.io/jnsgruk/zinc:0.4.9 --trust
```
2. Add integrations to oathkeeper:
```
juju relate oathkeeper zinc-k8s:auth-proxy
juju relate oathkeeper traefik-ingress:experimental-forward-auth
```
At this point, if zinc was not related to ingress, oathkeeper will get blocked with message `zinc-k8s is not related via ingress`.
After you relate zinc with ingress, oathkeeper will get active:
```
juju relate zinc-k8s traefik-ingress
```
3. Check the traefik routes:
```bash
juju ssh --container traefik traefik-ingress/0 bash
cd /opt/traefik/juju/
```
Zinc route should define and reference the middleware:
```yaml
# cat juju_ingress_ingress_6_zinc-k8s.yaml
http:
  middlewares:
    juju-sidecar-forward-auth-oathkeeper:
      forwardAuth:
        address: https://oathkeeper.test21.svc.cluster.local:4456/decisions
        authResponseHeaders:
        - X-User
    juju-sidecar-noprefix-test21-zinc-k8s:
      stripPrefix:
        forceSlash: false
        prefixes:
        - /test21-zinc-k8s
  routers:
    juju-test21-zinc-k8s-router:
      entryPoints:
      - web
      middlewares:
      - juju-sidecar-forward-auth-oathkeeper
      - juju-sidecar-noprefix-test21-zinc-k8s
      rule: PathPrefix(`/test21-zinc-k8s`)
      service: juju-test21-zinc-k8s-service
    juju-test21-zinc-k8s-router-tls:
      entryPoints:
      - websecure
      middlewares:
      - juju-sidecar-forward-auth-oathkeeper
      - juju-sidecar-noprefix-test21-zinc-k8s
      rule: PathPrefix(`/test21-zinc-k8s`)
      service: juju-test21-zinc-k8s-service
      tls: {}
  services:
    juju-test21-zinc-k8s-service:
      loadBalancer:
        servers:
        - url: http://zinc-k8s.test21.svc.cluster.local:4080
```

## Release Notes

- add `forward-auth` integration with oathkeeper
